File: schedule_generator.py
# schedule_generator.py
import sys
import argparse
import random
import csv
import json
from datetime import datetime, timedelta

class ScheduleGenerator:
    def __init__(self, days=5, periods=6, subjects=None, teachers=None, rooms=None,
                 start_time="09:00", duration=45):
        self.days = days
        self.periods = periods
        self.subjects = subjects or ["Math", "Physics", "Chemistry", "Biology", "History"]
        self.teachers = teachers or ["Smith", "Jones", "Williams", "Brown", "Davis"]
        self.rooms = rooms or ["101", "102", "103", "104", "105"]
        self.start_time = datetime.strptime(start_time, "%H:%M")
        self.duration = duration
        self.schedule = {}  # day -> list of (subject, teacher, room)

    def generate(self):
        # Ensure enough teachers and rooms for each period
        if len(self.teachers) < self.periods:
            print("Warning: Not enough teachers, adding dummy ones.")
            for i in range(self.periods - len(self.teachers)):
                self.teachers.append(f"T{i+1}")
        if len(self.rooms) < self.periods:
            print("Warning: Not enough rooms, adding dummy ones.")
            for i in range(self.periods - len(self.rooms)):
                self.rooms.append(f"R{i+1}")

        for day in range(self.days):
            day_schedule = []
            # Shuffle teachers and rooms for this day
            teachers_shuffled = random.sample(self.teachers, self.periods)
            rooms_shuffled = random.sample(self.rooms, self.periods)
            # Assign subjects randomly (allow repeats)
            subjects_shuffled = [random.choice(self.subjects) for _ in range(self.periods)]
            for p in range(self.periods):
                day_schedule.append((subjects_shuffled[p], teachers_shuffled[p], rooms_shuffled[p]))
            self.schedule[day] = day_schedule

    def display(self):
        print("\nGenerated Schedule:")
        print("-" * 80)
        print(f"{'Day':<6} {'Period':<8} {'Time':<14} {'Subject':<12} {'Teacher':<12} {'Room'}")
        print("-" * 80)
        for day in range(self.days):
            for period, (subject, teacher, room) in enumerate(self.schedule[day], start=1):
                start = self.start_time + timedelta(minutes=(period-1)*self.duration)
                end = start + timedelta(minutes=self.duration)
                time_str = f"{start.strftime('%H:%M')}-{end.strftime('%H:%M')}"
                print(f"{day+1:<6} {period:<8} {time_str:<14} {subject:<12} {teacher:<12} {room}")

    def export_csv(self, filename):
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            header = ["Day", "Period", "Time", "Subject", "Teacher", "Room"]
            writer.writerow(header)
            for day in range(self.days):
                for period, (subject, teacher, room) in enumerate(self.schedule[day], start=1):
                    start = self.start_time + timedelta(minutes=(period-1)*self.duration)
                    end = start + timedelta(minutes=self.duration)
                    time_str = f"{start.strftime('%H:%M')}-{end.strftime('%H:%M')}"
                    writer.writerow([day+1, period, time_str, subject, teacher, room])

    def export_json(self, filename):
        data = []
        for day in range(self.days):
            day_data = []
            for period, (subject, teacher, room) in enumerate(self.schedule[day], start=1):
                start = self.start_time + timedelta(minutes=(period-1)*self.duration)
                end = start + timedelta(minutes=self.duration)
                day_data.append({
                    "period": period,
                    "time": f"{start.strftime('%H:%M')}-{end.strftime('%H:%M')}",
                    "subject": subject,
                    "teacher": teacher,
                    "room": room
                })
            data.append({"day": day+1, "periods": day_data})
        with open(filename, 'w') as f:
            json.dump(data, f, indent=2)

def interactive():
    print("=== Schedule Generator ===")
    days = int(input("Enter number of days (1-7): ") or "5")
    periods = int(input("Enter number of periods per day (1-8): ") or "6")
    subjects = input("Enter subjects (comma-separated): ").split(',')
    subjects = [s.strip() for s in subjects if s.strip()]
    if not subjects:
        subjects = ["Math", "Physics", "Chemistry", "Biology", "History"]
    teachers = input("Enter teachers (comma-separated): ").split(',')
    teachers = [t.strip() for t in teachers if t.strip()]
    if not teachers:
        teachers = ["Smith", "Jones", "Williams", "Brown", "Davis"]
    rooms = input("Enter rooms (comma-separated): ").split(',')
    rooms = [r.strip() for r in rooms if r.strip()]
    if not rooms:
        rooms = ["101", "102", "103", "104", "105"]
    start_time = input("Start time (HH:MM, default 09:00): ") or "09:00"
    duration = int(input("Period duration (minutes, default 45): ") or "45")
    gen = ScheduleGenerator(days, periods, subjects, teachers, rooms, start_time, duration)
    gen.generate()
    gen.display()
    export = input("Export to file? (y/n): ").strip().lower()
    if export == 'y':
        fname = input("Filename (e.g., schedule.csv or schedule.json): ")
        if fname.endswith('.json'):
            gen.export_json(fname)
        else:
            gen.export_csv(fname)
        print(f"Saved to {fname}")

def cli():
    parser = argparse.ArgumentParser(description='Schedule Generator')
    parser.add_argument('--days', type=int, default=5, help='Number of days')
    parser.add_argument('--periods', type=int, default=6, help='Periods per day')
    parser.add_argument('--subjects', default='Math,Physics,Chemistry,Biology,History', help='Comma-separated subjects')
    parser.add_argument('--teachers', default='Smith,Jones,Williams,Brown,Davis', help='Comma-separated teachers')
    parser.add_argument('--rooms', default='101,102,103,104,105', help='Comma-separated rooms')
    parser.add_argument('--start', default='09:00', help='Start time HH:MM')
    parser.add_argument('--duration', type=int, default=45, help='Period duration in minutes')
    parser.add_argument('--output', help='Output filename (CSV or JSON)')
    args = parser.parse_args()
    subjects = [s.strip() for s in args.subjects.split(',')]
    teachers = [t.strip() for t in args.teachers.split(',')]
    rooms = [r.strip() for r in args.rooms.split(',')]
    gen = ScheduleGenerator(args.days, args.periods, subjects, teachers, rooms, args.start, args.duration)
    gen.generate()
    gen.display()
    if args.output:
        if args.output.endswith('.json'):
            gen.export_json(args.output)
        else:
            gen.export_csv(args.output)
        print(f"Saved to {args.output}")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        cli()
    else:
        interactive()

File: schedule_generator.rb
# schedule_generator.rb
require 'json'
require 'optparse'
require 'time'

class ScheduleGenerator
  attr_reader :days, :periods, :subjects, :teachers, :rooms, :start_time, :duration, :schedule

  def initialize(days: 5, periods: 6, subjects: nil, teachers: nil, rooms: nil, start_time: '09:00', duration: 45)
    @days = days
    @periods = periods
    @subjects = subjects || ['Math', 'Physics', 'Chemistry', 'Biology', 'History']
    @teachers = teachers || ['Smith', 'Jones', 'Williams', 'Brown', 'Davis']
    @rooms = rooms || ['101', '102', '103', '104', '105']
    @start_time = Time.parse(start_time)
    @duration = duration
    @schedule = {}
  end

  def generate
    if @teachers.size < @periods
      (@teachers.size...@periods).each { |i| @teachers << "T#{i+1}" }
    end
    if @rooms.size < @periods
      (@rooms.size...@periods).each { |i| @rooms << "R#{i+1}" }
    end
    @days.times do |day|
      day_schedule = []
      teachers_shuffled = @teachers.shuffle
      rooms_shuffled = @rooms.shuffle
      @periods.times do |p|
        subject = @subjects.sample
        day_schedule << {
          subject: subject,
          teacher: teachers_shuffled[p % teachers_shuffled.size],
          room: rooms_shuffled[p % rooms_shuffled.size]
        }
      end
      @schedule[day] = day_schedule
    end
  end

  def display
    puts "\nGenerated Schedule:"
    puts '-' * 80
    printf "%-6s %-8s %-14s %-12s %-12s %s\n", "Day", "Period", "Time", "Subject", "Teacher", "Room"
    puts '-' * 80
    @days.times do |day|
      @periods.times do |p|
        entry = @schedule[day][p]
        start_t = @start_time + p * @duration * 60
        end_t = start_t + @duration * 60
        time_str = "#{start_t.strftime('%H:%M')}-#{end_t.strftime('%H:%M')}"
        printf "%-6d %-8d %-14s %-12s %-12s %s\n", day+1, p+1, time_str, entry[:subject], entry[:teacher], entry[:room]
      end
    end
  end

  def export_csv(filename)
    CSV.open(filename, 'w') do |csv|
      csv << ['Day', 'Period', 'Time', 'Subject', 'Teacher', 'Room']
      @days.times do |day|
        @periods.times do |p|
          entry = @schedule[day][p]
          start_t = @start_time + p * @duration * 60
          end_t = start_t + @duration * 60
          time_str = "#{start_t.strftime('%H:%M')}-#{end_t.strftime('%H:%M')}"
          csv << [day+1, p+1, time_str, entry[:subject], entry[:teacher], entry[:room]]
        end
      end
    end
  end

  def export_json(filename)
    data = []
    @days.times do |day|
      periods_data = []
      @periods.times do |p|
        entry = @schedule[day][p]
        start_t = @start_time + p * @duration * 60
        end_t = start_t + @duration * 60
        time_str = "#{start_t.strftime('%H:%M')}-#{end_t.strftime('%H:%M')}"
        periods_data << {
          period: p+1,
          time: time_str,
          subject: entry[:subject],
          teacher: entry[:teacher],
          room: entry[:room]
        }
      end
      data << { day: day+1, periods: periods_data }
    end
    File.write(filename, JSON.pretty_generate(data))
  end
end

def interactive
  require 'csv'
  puts "=== Schedule Generator ==="
  print "Enter number of days (1-7): "
  days = gets.to_i
  days = 5 if days < 1 || days > 7
  print "Enter number of periods per day (1-8): "
  periods = gets.to_i
  periods = 6 if periods < 1 || periods > 8
  print "Enter subjects (comma-separated): "
  subjects = gets.chomp.split(',').map(&:strip)
  subjects = ['Math', 'Physics', 'Chemistry', 'Biology', 'History'] if subjects.empty?
  print "Enter teachers (comma-separated): "
  teachers = gets.chomp.split(',').map(&:strip)
  teachers = ['Smith', 'Jones', 'Williams', 'Brown', 'Davis'] if teachers.empty?
  print "Enter rooms (comma-separated): "
  rooms = gets.chomp.split(',').map(&:strip)
  rooms = ['101', '102', '103', '104', '105'] if rooms.empty?
  print "Start time (HH:MM, default 09:00): "
  start_time = gets.chomp
  start_time = '09:00' if start_time.empty?
  print "Period duration (minutes, default 45): "
  duration = gets.to_i
  duration = 45 if duration <= 0
  gen = ScheduleGenerator.new(days: days, periods: periods, subjects: subjects, teachers: teachers,
                              rooms: rooms, start_time: start_time, duration: duration)
  gen.generate
  gen.display
  print "Export to file? (y/n): "
  if gets.chomp.downcase == 'y'
    print "Filename (e.g., schedule.csv or schedule.json): "
    fname = gets.chomp
    if fname.end_with?('.json')
      gen.export_json(fname)
    else
      gen.export_csv(fname)
    end
    puts "Saved to #{fname}"
  end
end

def cli
  options = {}
  OptionParser.new do |opts|
    opts.on('--days N', Integer) { |v| options[:days] = v }
    opts.on('--periods N', Integer) { |v| options[:periods] = v }
    opts.on('--subjects S', String) { |v| options[:subjects] = v }
    opts.on('--teachers T', String) { |v| options[:teachers] = v }
    opts.on('--rooms R', String) { |v| options[:rooms] = v }
    opts.on('--start TIME', String) { |v| options[:start] = v }
    opts.on('--duration N', Integer) { |v| options[:duration] = v }
    opts.on('--output FILE', String) { |v| options[:output] = v }
  end.parse!
  days = options[:days] || 5
  periods = options[:periods] || 6
  subjects = options[:subjects] ? options[:subjects].split(',').map(&:strip) : nil
  teachers = options[:teachers] ? options[:teachers].split(',').map(&:strip) : nil
  rooms = options[:rooms] ? options[:rooms].split(',').map(&:strip) : nil
  start_time = options[:start] || '09:00'
  duration = options[:duration] || 45
  output = options[:output]
  gen = ScheduleGenerator.new(days: days, periods: periods, subjects: subjects, teachers: teachers,
                              rooms: rooms, start_time: start_time, duration: duration)
  gen.generate
  gen.display
  if output
    if output.end_with?('.json')
      gen.export_json(output)
    else
      require 'csv'
      gen.export_csv(output)
    end
    puts "Saved to #{output}"
  end
end

if ARGV.empty?
  interactive
else
  cli
end
